### ISSUE TITLE
DROID matching file against container signature specification where it shouldn't

### ISSUE THREAD

**Issue by ross-spencer** [digital-preservation/droid#105]
This was spotted by @richardlehane in one of my Govdocs trials, and I've verified it looks like a bug. 

The file 644692.xls in the Govdocs Select corpus is being matched by the container matching engine as fmt/61 

DROID 6.2.1, MaxBytesToScan = -1

If I look inside the OLE2 I see that a 'Workbook' stream doesn't exist but a 'Book' does, and so DROID must be matching against container sequence '2020'

      <ContainerSignature Id="2020" ContainerType="OLE2">
         <Description>Microsoft Excel 97 OLE2</Description>
         <Files>
             <File>
                 <Path>Book</Path>
                 <BinarySignatures>
                     <InternalSignatureCollection>
                        <InternalSignature ID="308">
                            <ByteSequence Reference="BOFoffset">
                                <SubSequence Position="1" SubSeqMinOffset="0" SubSeqMaxOffset="0">
                                    <Sequence>09 08</Sequence>
                                </SubSequence>
                                <SubSequence Position="2" SubSeqMinOffset="0" SubSeqMaxOffset="4">
                                    <Sequence>00 06 05 00</Sequence>
                                </SubSequence>
                            </ByteSequence>
                        </InternalSignature>
                    </InternalSignatureCollection>
                 </BinarySignatures>
             </File>
         </Files>
      </ContainerSignature>

Bytes 09 08 are in the correct place. Bytes 00 06 05 00 are matched further down the file however than max offset 4. 

This screenshot shows their first appearance at decimal offset 77169: 

![image](https://cloud.githubusercontent.com/assets/1880412/17653614/561644a4-62ec-11e6-9d69-4b6ecb50b451.png)

Attached is a zip file containing the file for you. It would be helpful if you can verify the behavior. If verified then it seems that the container matching engine isn't quite working as expected. 

(**n.b.** If it helps, the issue also looks to be present in 6.1.5 which iirc was before the last Byteseek upgrade, and so might focus the search for the bug)

[644692.zip](https://github.com/digital-preservation/droid/files/417427/644692.zip)

**Comment by snail1966** [comment]
I'm seeing the fmt/59 identification when run with the default maximum bytes to scan setting (65536).  And fmt/61 when this is set to -1.  Same behaviour in earlier version 6.1.5

**Comment by snail1966** [comment]
To note for completeness - with the command line DROID, the identification is fmt/59 if run with just a binary signature, and fmt/61 if you add a container signature (-Nc option).  This is using binary signature version 86 and container signature 20160727.

**Comment by Dclipsham** [comment]
Was hoping this would have gone away with recent work on container/binary signature parsing, but this remains an issue in 6.5 testing so paging @nishihatapalmer for thoughts

File [https://github.com/digital-preservation/droid/files/417427/644692.zip] is an XLS that conforms to container signature for fmt/59 - ID 2000 ('Book' file, then we want to seek 0908{2}0050500 but Container sigs don't seem to like wildcards) - We've expressed this instead with:
`    <ContainerSignature Id="2000" ContainerType="OLE2">
      <Description>Microsoft Excel 5.0/95 OLE2</Description>
      <Files>
        <File>
          <Path>Book</Path>
          <BinarySignatures>
            <InternalSignatureCollection>
              <InternalSignature ID="2000">
                <ByteSequence Reference="BOFoffset">
                  <SubSequence Position="1" SubSeqMinOffset="0"
                    SubSeqMaxOffset="0">
                    <Sequence>09 08</Sequence>
                  </SubSequence>
                  <SubSequence Position="2" SubSeqMinOffset="0"
                    SubSeqMaxOffset="4">
                    <Sequence>00 05 05 00</Sequence>
                  </SubSequence>
                </ByteSequence>
              </InternalSignature>
            </InternalSignatureCollection>
          </BinarySignatures>
        </File>
      </Files>
    </ContainerSignature>`
Intending this as us wanting 0908 from offset 0, 00050500 from offset 4. This does identify correctly in isolation. 

_However_, fmt/61 has the following:
`    <ContainerSignature Id="2020" ContainerType="OLE2">
      <Description>Microsoft Excel 97 OLE2</Description>
      <Files>
        <File>
          <Path>Book</Path>
          <BinarySignatures>
            <InternalSignatureCollection>
              <InternalSignature ID="2020">
                <ByteSequence Reference="BOFoffset">
                  <SubSequence Position="1" SubSeqMinOffset="0"
                    SubSeqMaxOffset="0">
                    <Sequence>09 08</Sequence>
                  </SubSequence>
                  <SubSequence Position="2" SubSeqMinOffset="0"
                    SubSeqMaxOffset="4">
                    <Sequence>00 06 05 00</Sequence>
                  </SubSequence>
                </ByteSequence>
              </InternalSignature>
            </InternalSignatureCollection>
          </BinarySignatures>
        </File>
      </Files>
    </ContainerSignature>`
Basically intended to represent 0908{2}00060500

For some reason however, DROID is returning a match against fmt/61, but not against fmt/59. If we comment out the relevant signature for fmt/61 (id="2020") then we *do* get a match on fmt/59

So, what could be happening in the above case to cause this behaviour? And perhaps more usefully:

How to properly represent wildcards within containers? (container-specific syntax for wildcards is absent here - https://github.com/digital-preservation/droid/blob/master/Signature%20syntax.md)

cc @jcharlet @sparkhi

**Comment by nishihatapalmer** [comment]
Quite a lot to unpack here.  

**Container Signature Wildcards**
Container signatures in DROID 6.5 will support the full wildcard PRONOM syntax directly.  Prior to 6.5, these have to be represented by splitting into XML manually, like the binary signatures, which is tricky to do by hand.  DROID 6.5 does also have a tool (sigtool) which can output this for you (allowing you to more easily create signatures for older versions of DROID).

**Use of SubSequence**
If a ByteSequence has more than one SubSequence, they are always separated from each other using a * wildcard implicitly.  So it's no surprise that the second subsequence is sometimes found much further away than the subseqmin/max offsets.  The subseqmin / max offsets don't do what you expect them to here.  

If you don't want a * wildcard search - just a fixed or variable gap implied by the subseqmin/max offsets, these should be modelled as either Left or Right Fragments within a single SubSequence.  The longest part of the subsequence should be the main sequence of the subsequence.  So for `0908{2}0050500` we could have:

```
  <Description>Microsoft Excel 5.0/95 OLE2</Description>

  <Files>

    <File>

      <Path>Book</Path>

      <BinarySignatures>

        <InternalSignatureCollection>

          <InternalSignature ID="2000">

            <ByteSequence Reference="BOFoffset">

              <SubSequence Position="1" SubSeqMinOffset="0" SubSeqMaxOffset="0">

                <Sequence>00 05 05 00</Sequence>

                <LeftFragment Position="1" MinOffset="2" MaxOffset="2">09 08</LeftFragment>

              </SubSequence>

            </ByteSequence>

          </InternalSignature>

        </InternalSignatureCollection>

      </BinarySignatures>

    </File>

  </Files>

</ContainerSignature>
```

**Comment by nishihatapalmer** [comment]
A general observation - it's not only Container signatures that don't like wildcards.  Binary signatures can't represent wildcards directly either.  It's all done with fragments and subsequences and complicated XML.  At least, they couldn't until 6.5.

**Comment by nishihatapalmer** [comment]
Not sure why disabling fmt/61 causes fmt/59 to match.  Container matching is a little different to binary matching.

I can have a look if that's a burning issue - but I think the main issue of matching will be resolved by using fragments rather than subsequences to model fixed or limited gaps (rather than * wildcards)

**Comment by Dclipsham** [comment]
Ahh no, I've spotted it. fmt/61 has priority set over fmt/59. I don't believe it should have though. That's really helpful thanks Matt. I'll test that today.
